### DISCUSSION TITLE
Better generic entities for testing

### DISCUSSION THREAD

**Discussion by adapap** [StevensSEC/pokemonbattlelib#280] [Ideas]
As #278 reveals, adding new components to the library usually reveals the failure of previously stable tests as a result of additional functionality that invalidates previous test code. I think a good way to avoid this problem is to create null-like values (that function as no-op entities) to ensure that the focus of a test is on one thing and one thing only.

An example might be a type-less, invincible, generic Pokemon that can be customized to meet certain needs of tests by adding respective properties to the field.

```go
GenericPokemon = Pokemon{}
GenericPokemon.metadata[MetaGeneric] = true
// ...
GenericPokemon.Type = TypeFire
GenericPokemon.HP = 120
```

The cases in which this kind of style would be applicable to includes:
- Testing move damage without PP, abilities, or other side effects
- Simulating rounds _with_ or _without_ fainting predictably
- Avoiding STAB/weakness matchups from moves and elemental types
- Accidentally triggering abilities based on coincidental setups, such as weather or status effects

I'm not too positive how we could approach this task, but I think limiting the use of `GeneratePokemon` in tests is a good start.

**Comment by dyc3** [comment]
I don't particularly like the idea of a metadata flag to solve this problem, sounds like a large and messy addition to the battle logic. I think a better way to go about it is to change how the pokedex works. I briefly touched on the idea during one of the GBMs.

Here's my idea:
Instead of the pokedex just simply containing the data, the pokedex could act more like a registry. 

This is what I would imagine the API would look like:
```go
RegisterPokemon(d PokemonData)
RegisterMove(d MoveData)
RegisterItem(d ItemData)
etc...
```
These functions would basically just add the data to their respective lists, but we can do preprocessing/validation to the data. For example, they should automatically assign IDs, dex numbers, etc. in an incrementing fashion, in the order that the `Register` methods were called.

We could also include build flags to exclude the auto generated data from the build, and allow the user to supply their own pokemon, moves, items, etc. 

This would solve our problem by allowing us to invent the pokemon we need that meets certain criteria.

-----

Another possible solution would be to make a new generator to generate a pokemon that meets a certain set of rules.

```go
GenerateAnyPokemon(WithAbility(AbilityGuts))
GenerateAnyPokemon(WithType(TypeWater), WithMoves(SuperEffectiveAgainst(TypeGrass)))
GenerateAnyPokemon(WithType(TypeWater), NotWithMoves(SuperEffectiveAgainst(TypeGrass)))
GenerateAnyPokemon(NotWithMoves(WithType(TypeFire)))
```

-----

Additionally, (just kinda spitballing) we could make gomega matchers to assert that certain things aren't happening. Maybe something similar to the pokemon validation rules. Maybe something we could put in `AfterEach`?

```go
Expect(transactions).ToNot(UseBattleRule(TypeMatchup | Abilities))
Expect(transactions).ToNot(UseBattleRule(AllDamageRules ^ TypeMatchup))
```

Or maybe it would be better to validate the battle itself with these rules in `JustBeforeEach`?
